Gift exchange added!

zarutian opened the pull request:
So people can pass various things between themselfs for purposes of making complex examples.

ps. I hope difference of coding style is not too clashing.

zarutian committed:
Gift exchange added!

so people can pass stuff around.

dckc commented:
I don't see any reason to use a completely different style from the rest of the code:

> Monte’s blocks are usually indentation-delimited: ...
> Monte code should always uses four spaces for each indentation level.
 -- http://monte.readthedocs.io/en/latest/symbols.html#brackets-indentation-and-blocks

dckc commented on airbrus.mt:
```diff
@@ -68,6 +68,73 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+    object GEI {
+      """Your interface to the giftExchange"""
+      to give(recipiant :Str, giftName :Str, gift :Any) :Void {
+        """signiture: give(recipiant :Str, giftName :Str, gift :Any) :Void"""
```
typo: signature

dckc commented on airbrus.mt:
```diff
@@ -68,6 +68,73 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+  
+  def give(giver :Str, recipiant :Str, giftName :Str, gift :Any) :Void {
+    def giftBox := gift  # Allways wrap your presents! (For when gifts are promises)
+    if (wanters.fetch(recipiant, pass_duck) != duck) {
```
typo: recipient

zarutian committed:
Update airbrus.mt

restyled the addition.

dckc commented on airbrus.mt:
```diff
@@ -68,6 +68,62 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+    def pass_duck () :Any :
+        return duck
+        
+    def give(giver :Str, recipiant :Str, giftName :Str, gift :Any) :Void :
```
typo: recipient

dckc commented on airbrus.mt:
```diff
@@ -68,6 +68,62 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+  
+    def accept(recipiant :Str, giver :Str, giftName :Str) :Any :
+        def promres := Ref.make_promise()
+        var [promise, resolver] := promres
```
consider a more concise:
```
    def [var promise, var resolver] := Ref.make_promise()
```

Do they need to be var? Let's see...

dckc commented on airbrus.mt:
```diff
@@ -68,6 +68,62 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+            to give(recipiant :Str, giftName :Str, gift :Any) :Void :
+                """signiture: give(recipiant :Str, giftName :Str, gift :Any) :Void"""
+                give(nickname, recipiant, giftName, gift)
+                return
```
What is this `return` for?

zarutian committed:
Fixed a typo!

zarutian committed:
Changed my mind on ephimeriality of gifts.

dckc commented on airbrus.mt:
```diff
@@ -68,6 +68,62 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+        object GEI:
+            """Your interface to the giftExchange"""
+            to give(acceptor :Str, giftName :Str, gift :Any) :Void :
+                """signiture: give(acceptor :Str, giftName :Str, gift :Any) :Void"""
```
typo: signature

dckc commented on airbrus.mt:
```diff
@@ -188,7 +244,7 @@ def main(argv, => Timer,
                 def instanceEnv := ["&&help" => &&brusHelp]
                 def userEnv := userEnvironments.fetch(user.getNick(),
-                                                      fn {baseEnv | instanceEnv})
+                                                      fn { return baseEnv | instanceEnv | ["giftExchange" => make_giftExchange_interface(user.getNick())] })
```
why add `return` in the `fn`? it's implicit.

zarutian replied:
been bitten too many times in various other programming languages.

It is also a holdover from when I was playing with E. (Which does not have implict return iirc)

MostAwesomeDude replied:
This is a bug; `return` is not legal inside `fn` on its own, and the code will end up calling the (stale) `__return` ejector from `main`'s frame.

zarutian committed:
Got rid of the duck.

zarutian committed:
moved creation of an statefull object.

MostAwesomeDude commented on airbrus.mt:
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
 def makeLineTube() as DeepFrozen:
     return makePumpTube(makeSplitPump(b`$\n`))
 
+def make_giftExchange() :Any as DeepFrozen:
```
Style nit: `makeGiftExchange`

zarutian replied:
snit fixed.

MostAwesomeDude commented on airbrus.mt:
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+    def wanters := [].asMap().diverge()
+        
+    def give(giver :Str, acceptor :Str, giftName :Str, gift :Any) :Void :
+        def giftBox := gift  # Allways wrap your presents! (For when gifts are promises)
```
This does nothing. If you want to ensure that `gift` is near, guard it as `gift :Near`.

MostAwesomeDude commented on airbrus.mt:
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+        
+    def give(giver :Str, acceptor :Str, giftName :Str, gift :Any) :Void :
+        def giftBox := gift  # Allways wrap your presents! (For when gifts are promises)
+        if (wanters.contains(acceptor):
```
Syntax error; missing closing parenthesis.

zarutian replied:
closing parenthesis added.

MostAwesomeDude commented on airbrus.mt:
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+        if (wanters.contains(acceptor):
+            def wants := wanters[acceptor]
+            if (wants.contains(giftName)):
+              wants.fetch(giftName)[1].resolve(&giftBox)
```
This `&giftBox` can just be `&gift`. So can the other one.

zarutian replied:
Changed giftBox to be the result of _makeFinalSlot(). This is needed for when gifts are themselfs promises.

MostAwesomeDude commented on airbrus.mt:
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+        return
+  
+    def accept(acceptor :Str, giver :Str, giftName :Str) :Any :
+        def promres := Ref.make_promise()
```
Did you mean `Ref.promise()`?

zarutian replied:
Yes, fixed.

MostAwesomeDude commented on airbrus.mt:
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+  
+    def accept(acceptor :Str, giver :Str, giftName :Str) :Any :
+        def promres := Ref.make_promise()
+        var [promise, resolver] := promres
```
Should be folded with the previous line. Also `var` here doesn't actually create any mutable names; did you want `def [var promise, var resolver] := promres`? Also, for style, these names are traditionally `p` and `r`.

MostAwesomeDude commented on airbrus.mt:
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+            givers[giver] := [].asMap().diverge()
+        def gifts := givers[giver]
+        gifts[giftName] := &giftBox
+        return
```
Not needed.

zarutian replied:
Implict returns of control flow rub me the wrong way. Harmless to leave in.

MostAwesomeDude commented on airbrus.mt:
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+        if (givers.contains(giver)):
+            def gifts := givers[giver]
+            if (gifts.contains(giftName)):
+                resolver.resolve(gifts.fetch(giftName)
```
Syntax error.

zarutian replied:
yet another closing parenthesis missing. Fixed.

MostAwesomeDude commented on airbrus.mt:
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+        if (wanters.contains(acceptor)):
+            def wants := wanters[acceptor]
+            if (wants.contains(giftName)):
+                [promise, resolver] = wants.fetch(giftName)
```
This assignment isn't legal.

MostAwesomeDude commented on airbrus.mt:
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+            wanters.set(acceptor, [].asMap().diverge())
+        def wants := wanters.fetch(acceptor)
+        wants.set(giftName, promres)
+        return promise
```
Since all branches of control flow return `promise`, consider refactoring to have fewer returns.

MostAwesomeDude commented on airbrus.mt:
```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
+    def givers  := [].asMap().diverge()
+    def wanters := [].asMap().diverge()
+        
+    def give(giver :Str, acceptor :Str, giftName :Str, gift :Any) :Void :
```
Shouldn't this function and the following `accept` function be rolled into the methods which call them further down in the code?

zarutian replied:
Rolled them in. Started out this way due to habit of writing modules in Lua in this style.

MostAwesomeDude reviewed (changes requested):
I found some issues.

zarutian committed:
Fixing some of the issues @MostAwesomeDude pointed out.

zarutian committed:
Ditto.

zarutian committed:
Done for now.

zarutian committed:
Now, definitly done for now.

zarutian committed:
Updated airbrus.mt w/ hope it passes review.

zarutian commented:
I request a review of changes done so far.

zarutian committed:
Fixed a syntax error.

Apearantly, Monte does not have destructuring assignment. I see I have been spoilt by Lua.
